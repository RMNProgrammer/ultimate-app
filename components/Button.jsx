import React from 'react'
import { css, cx } from '@emotion/css'
import { useTheme } from '@emotion/react'

const Button = ({ type = 'button', children }) => { 
   const theme = useTheme()

   return(
      <div 
         className={css`
            width: 100%;
         `}
      >
         <button  type={type} 
                  className={css`
                     padding:15px;
                     font-size:14px;
                     cursor:pointer;
                     border-radius:20px;
                     background-color:white;
                     color:${theme.colors.primary};
                     border:2px solid ${theme.colors.secondary};
                  `}
         >
            {children}
         </button>
      </div>
   )
}

export default Button