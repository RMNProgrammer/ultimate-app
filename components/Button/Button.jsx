import React from 'react'
import { css, cx } from '@emotion/css'
import { useTheme } from '@emotion/react'

const Button = ({ type = 'button', children, onClick = () => {}, loading }) => { 
   const theme = useTheme()

   return(
      <div 
         className={css`
            width: 100%;
         `}
      >
         <button  type={type} 
                  className={css`
                     color:white;
                     width: 100%;
                     cursor:pointer;
                     position: relative;
                     display:inline-block;
                     text-decoration: none;
                     transition-duration: 0.4s;
                     padding:${theme.spaces[4]};
                     border:${theme.borderRadius[1]};
                     border-radius:${theme.borderRadius[1]};
                     background-color:${theme.colors.primary};
                     font-size:${theme.typography.paragraph.fontSize};
                        &:hover{
                           box-shadow: ${theme.boxShadow.light}
                        }
                     }
                  `}
                  onClick={onClick}
         >
            { loading && 'چند لحظه صبر کنید...' }
            { !loading && children }
         </button>
      </div>
   )
}

export default Button