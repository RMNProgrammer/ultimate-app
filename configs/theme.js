const theme = {
    colors: {
        primary: '#f03e3e',
        backgroundColorPrimary: 'rgba(220, 221, 225,0.5)',
        backgroundColorDefault: '#f8f9fa',
    },
    typography: {
        h1: {
            fontSize: 1.502,
            lineHeight: 1.2,
            fontWeight: 500,
        },
        h2: {
            fontSize: 1.352,
            lineHeight: 1.2,
            fontWeight: 500,
        },
        h3: {
            fontSize: 1.174,
            lineHeight: 1.2,
            fontWeight: 500,
        },
        h4: {
            fontSize: 1.016,
            lineHeight: 1.2,
            fontWeight: 500,
        },
        h5: {
            fontSize: 0.875,
            lineHeight: 1.2,
            fontWeight: 500,
        },
        paragraph: {
            fontSize: 1.100,
            lineHeight: 1.2,
            fontWeight: 500,
        },
        small: {
            fontSize: 1.352,
            lineHeight: 1.2,
            fontWeight: 100,
        },
    },
    spaces: ['1px','2px','4px','8px','16px','32px'],
    borderRadius: ['1px','2px'],
    border: ['1px','2px'],
    borderColor: {
        light: '#e3e4e6',
    },
    boxShadow: {
        light: '0 4px 16px rgba(0, 1, 31, 0.2)'
    },
}

export default theme