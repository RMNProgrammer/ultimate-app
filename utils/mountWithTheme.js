import React from 'react'
import theme from '../configs/theme'
import renderer from 'react-test-renderer'
import { ThemeProvider } from '@emotion/react'

const mountWithTheme = (children) => {
   return renderer.create(
      <ThemeProvider theme={theme}>
         {children}
      </ThemeProvider>
   )
}

export default mountWithTheme
