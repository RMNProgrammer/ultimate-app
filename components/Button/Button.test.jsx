import React from 'react'
import Button from './Button'
import theme from '../../configs/theme'
import { ThemeProvider } from '@emotion/react'
import mountWithTheme from '../../utils/mountWithTheme'
import { fireEvent, render } from '@testing-library/react'

describe('Button', () => {
   test('Renders Correctly', () => {
      const component = mountWithTheme(<Button>Sample Button</Button>)
      const tree = component.toJSON()
      expect(tree).toMatchSnapshot()
   })

   test('Shows Loading state', () => {
      const component = mountWithTheme(
         <Button loading>Sample Button with loading</Button>
      )
      const tree = component.toJSON()
      expect(tree).toMatchSnapshot()
   })

   test('Renders in DOM correctly', () => {
      const { getByText } = render(
         <ThemeProvider theme={theme}>
            <Button>Sample Button in DOM</Button> 
         </ThemeProvider>
      )
      expect(getByText(/Button/)).toBeTruthy()
   })

   test('OnClick works fine', () => {
      let counter = 0
      const { getByText } = render(
         <ThemeProvider theme={theme}>
            <Button onClick={() => { counter += 1}}>
               Sample Button in DOM
            </Button> 
         </ThemeProvider>
      )
      fireEvent.click(getByText(/Button/))
      expect(counter).toBe(1)
   })
})
