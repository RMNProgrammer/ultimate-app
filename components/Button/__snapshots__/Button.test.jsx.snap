// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button Renders Correctly 1`] = `
<div
  className="css-wemwrz"
>
  <button
    className="css-16ru7d"
    onClick={[Function]}
    type="button"
  >
    Sample Button
  </button>
</div>
`;

exports[`Button Shows Loading state 1`] = `
<div
  className="css-wemwrz"
>
  <button
    className="css-16ru7d"
    onClick={[Function]}
    type="button"
  >
    چند لحظه صبر کنید...
  </button>
</div>
`;
